### DISCUSSION TITLE
Unable to set custom session id

### DISCUSSION THREAD

**Discussion by deepsidhu1313** [symfony/symfony#50583] [Q&A]
To overcome thirdparty cookie issue, i am trying to set a custom session id for different users using `Symfony\Component\HttpFoundation\Session\Session` class, for some reason symfony generates a new id instead of the id i am trying to pass. Does anyone know if i am doing anything wrong ? if i change the sequence then it throws exception that session has already started.
```
$this->session->setId($session_id);
$this->session->start();
```
(could anyone confirm whether this could be a php bug, i am using php 8.1)

**Comment by MatTheCat** [comment]
Sorry if I don’t answer your question directly but this looks like a X/Y problem to me. What is the “thirdparty cookie issue”? Would changing the session’s name fix it?

**Reply by deepsidhu1313** [reply]
Thanks for your response. 🙂 
in my case it is an app iframed on a platform. When user visit that platform in incognito or private mode, third party cookies get disabled. Since it is an iframe and domain doesn't match the platform domain browser will consider all cookies from my app as third-party. To overcome this we rely on platform to provide us a session id, which theoretically we should set in our code. Just to be clear it is not session name, it is ID (🤔 if they are not the same thing behind the scenes).

**Reply by MatTheCat** [reply]
Soo I did some investigation and found #45186 but this was never implemented following https://github.com/symfony/symfony/pull/47071#pullrequestreview-1051922328.

Problem is, sessions handlers are wrapped into a `SessionHandlerProxy`, so your custom handler would need to extend `AbstractProxy` and implement `SessionIdInterface`.

EDIT: opened #50626

**Reply by deepsidhu1313** [reply]
Thanks for your reply, i will give it a try

**Reply by deepsidhu1313** [reply]
Not sure how it helps setting the a custom session id using `setId($some_id)` function of `Symfony\Component\HttpFoundation\Session\Session` ?? Shouldn't we change 

```php
        if ($sessionId && $this->saveHandler instanceof AbstractProxy && 'files' === $this->saveHandler->getSaveHandlerName() && !preg_match('/^[a-zA-Z0-9,-]{22,250}$/', $sessionId)) {
            // the session ID in the header is invalid, create a new one
            session_id(session_create_id());
        }
```

to 

```php
if ($sessionId && $this->saveHandler instanceof AbstractProxy && 'files' === $this->saveHandler->getSaveHandlerName() && !preg_match('/^[a-zA-Z0-9,-]{22,250}$/', $sessionId)) {
            if(!empty($sessionId = $this->getId())) {
            session_id($sessionId);
            } else {
            // the session ID in the header is invalid, create a new one
            session_id(session_create_id());
            }
        }

```

in `NativeSessionStorage` class

**Reply by MatTheCat** [reply]
You cannot use `Session::setId` yourself because Symfony forces the [`session.use_strict_mode`](https://www.php.net/manual/en/session.configuration.php#ini.session.use-strict-mode) ([which is a good thing](https://blog.ohgaki.net/risk-of-the-session-adoption)), which rejects preset ids if they don’t match an initialized session.

**Reply by MatTheCat** [reply]
From what I understood you need to create a class extending `SessionHandlerProxy` and implementing `SessionIdInterface`:

```php
<?php

namespace App\Session;

use Symfony\Component\HttpFoundation\Session\Storage\Proxy\SessionHandlerProxy;

class CustomIdSessionHandler extends SessionHandlerProxy implements \SessionIdInterface
{
    public function create_sid(): string
    {
        return 'a-valid-session-identifier';
    }
}
```

Then you must register it as a service and wire its `$handler` argument to the service corresponding to your `framework.session.handler_id` (this is `session.handler.native` by default):

```yaml
services:
    App\Session\CustomIdSessionHandler:
        arguments:
            - '@session.handler.native'
```

At last, you must replace `framework.session.handler_id` accordingly to your service’s id:

```yaml
framework:
    session:
        handler_id: App\Session\CustomIdSessionHandler
```

I’m not sure this is the way to go yet, but it works.
